src/Components/Popover/Popover.tsx

```diff
@@ -0,0 +1,476 @@
+  /** Type of interaction to show the popover dialog */
+  showEvent: PopoverInteraction
+  /** Type of interaction to hide the popover dialog */
+  hideEvent: PopoverInteraction
+  /** Pixel distance between trigger and popover dialog */
+  distanceFromTrigger: number
+  /** Where to position the popover relative to the trigger (assuming it fits there) */
+  position: PopoverPosition
+  /** Means of applying color to popover */
+  type: PopoverType
+  /** Renders the popover dialog visible initially */
+  initiallyVisible: boolean
+}
+
+interface State {
+  expanded: boolean
+  triggerRect: ClientRect | DOMRect | null
+}
+
+interface PopoverFlush {
+  first: boolean
+  last: boolean
+}
+
+export class Popover extends Component<Props, State> {
+  public static readonly displayName = 'Popover'
+
+  public static defaultProps = {
+    color: ComponentColor.Primary,
+    testID: 'popover',
+    showEvent: PopoverInteraction.Click,
+    hideEvent: PopoverInteraction.Click,
+    distanceFromTrigger: 4,
+    position: PopoverPosition.Below,
+    type: PopoverType.Solid,
+    initiallyVisible: false,
+  }
+
+  public static DismissButton = PopoverDismissButton
+
+  private triggerRef = createRef<HTMLDivElement>()
+  private dialogRef = createRef<HTMLDivElement>()
+
+  constructor(props: Props) {
+    super(props)
+
+    this.state = {
+      expanded: this.props.initiallyVisible,
+      triggerRect: null,
+    }
+  }
+
+  public componentDidMount() {
+    if (this.props.initiallyVisible) {
+      this.handleShowDialog()
+    } else {
+      this.handleUpdateTriggerRect()
+    }
+  }
+
+  public render() {
+    const {testID, children, id} = this.props
+
+    return (
+      <ClickOutside onClickOutside={this.handleClickOutside}>
+        <div
+          className={this.className}
+          data-testid={testID}
+          id={id}
+          onClick={this.handleTriggerClick}
+          onMouseOver={this.handleTriggerMouseOver}
+          onMouseOut={this.handleTriggerMouseOut}
+          ref={this.triggerRef}
+        >
+          {children}
+          {this.dialog}
+        </div>
+      </ClickOutside>
+    )
+  }
+
+  private get className(): string {
+    const {color, className, type} = this.props
+
+    return classnames('cf-popover', {
+      [`cf-popover__${color}`]: color,
+      [`${className}`]: className,
+      [`cf-popover__${type}`]: type,
+    })
+  }
+
+  private get dialog(): JSX.Element {
+    const {triggerRect, expanded} = this.state
+    const {contents, distanceFromTrigger} = this.props
+    const dialogPosition = this.calculateDialogPosition()
+    let dialogStyles: CSSProperties = {
+      visibility: expanded ? 'visible' : 'hidden',
+    }
+    let caretStyles: CSSProperties = {}
+
+    if (triggerRect && this.dialogRef.current) {
+      const dialogRect = this.dialogRef.current.getBoundingClientRect()
+      let dialogFlush
+
+      switch (dialogPosition) {
+        case PopoverPosition.Above:
```

Comment:
Might be helpful to extract this to a pure function or composition of pure functions. This logic is gnarly and I know it has to be, but maybe we can prevent it from become gnarlier by explicitly declaring the dependencies of this routine as the arguments to a function.

Otherwise it can access everything on `this`, which is far more than it needs. I think over time that leads to tight coupling / people trying to do too much all in one place.